SSA/Projects/InstCombine/AliveAutoGenerated.lean

```diff
@@ -1,12 +1,101 @@
 import SSA.Projects.InstCombine.LLVM.EDSL
-import SSA.Projects.InstCombine.Tactic
+import SSA.Projects.InstCombine.AliveStatements
+import Mathlib.Tactic
 open MLIR AST
 
+namespace AliveAutoGenerated
+
 abbrev ICom.Refinement (src tgt : Com (φ:=0) Γ t) (h : Goedel.toType t = Option α := by rfl) : Prop :=
   ∀ Γv, Bitvec.Refinement (h ▸ src.denote Γv) (h ▸ tgt.denote Γv)
 
 infixr:90 " ⊑ "  => ICom.Refinement
 
+/- TODO: is this theorem true? -/
+@[simp]
+theorem Bitvec.ofNat_zero : Bitvec.ofNat w 0 = 0 := by
+  induction w
+  case zero => simp[Bitvec.ofNat]
+  case succ w' wh =>
+    simp[Bitvec.ofNat]
+    rw[wh]
+    simp[OfNat.ofNat, Zero.zero]
+    sorry
+
+@[simp]
+theorem one_by_two_is_zero : 1 / 2 = (0 : Nat) := rfl
+
+@[simp]
+theorem Bitvec.ofNat_one : Bitvec.ofNat w 1 = 1 := by
+  cases w
+  case zero => simp[Bitvec.ofNat]
+  case succ w' =>
+    simp[Bitvec.ofNat]
+    sorry -- to be proven...
+
+
+
+/--
+- We first generalize `v0...vn`, which gives us names for the accessors of the context.
+- Then we `cases` on these, to clear away the case where the variables could be `none`, leaving us with the
+  interesting case where the variable is a `some value`.
+- At this point, the proof state is what we want. However, the names `x0..xn` are *inaccessible* to the
+  user since the names were generated from within the macro.
+- So, we choose to revert the variables `x0...xn` so that the user can introduce them as she likes.
+- If we directly revert `x0...xn`, we wind up getting a proof state of the form
+    `∀ x0 V0 ... xn Vn`, where the `{ Vi }` is pure noise.
+- So we first clear the `V0...Vn` equations, and then revert `x0...xn`.
+- This gives us a statement of the form `∀ x0... xn, GOAL`
+-/
+macro "simp_mlir" : tactic =>
+  `(tactic|
+      (
+      dsimp only[]
+      unfold ICom.Refinement
+      intros Γv
+      try simp only[ICom.denote, IExpr.denote, HVector.denote, OpDenote.denote, InstCombine.Op.denote, HVector.toPair, pairMapM,
+        HVector.toTuple]
+      try simp[bind, Option.bind, pure]
+      try simp[DerivedContext.ofContext, DerivedContext.snoc, Ctxt.snoc] -- cannot rewrite with simp theorem 'motive is not type correct'
+      try simp[MOp.instantiateCom, InstCombine.MTy.instantiate, ConcreteOrMVar.instantiate,
+          Vector.get, HVector.toSingle, HVector.toTuple]
+      try simp[List.nthLe]
```

Comment:
```suggestion
      simp [ICom.denote, IExpr.denote, HVector.denote, OpDenote.denote, InstCombine.Op.denote,
        HVector.toPair, pairMapM, HVector.toTuple, bind, Option.bind, pure,
        DerivedContext.ofContext, DerivedContext.snoc, Ctxt.snoc,
        MOp.instantiateCom, InstCombine.MTy.instantiate, ConcreteOrMVar.instantiate,
        Vector.get, HVector.toSingle, HVector.toTuple, List.nthLe
      ]
```
Can this be just a single (appropriately squeezed) simp call?

Reply:
Done!